dashboard/app/models/user.rb

```diff
@@ -289,7 +289,7 @@ def preprocess_school_info(school_info_attr)
   # old school info object doesn't have a NCES school ID associated with it
   # @param new_school_info a school_info object to compare to the user current school information.
   def update_school_info(new_school_info)
-    if school_info.try(&:school).nil? || new_school_info.try(&:school)
+    if (school_info.try(&:school).nil? || new_school_info.try(&:school)) && new_school_info.complete?
       self.school_info_id = new_school_info.id
```

Comment:
Aren't we trying to deprecate `User.school_info_id`?  I understand we're leaning on this and the after_save hook, but I'd be happier if we could call something like `amend_school_history(new_school_info)` that made it more obvious what's going on here.

Reply:
We are? What's the plan to deprecate `school_info_id` but keep that association?

Reply:
I remember we had a discussion about this.

Reply:
I'm sure we did I've just forgotten :)

Reply:
That was my assumption. It seems redundant, and maybe misleading to keep `belongs_to :school_info` around when User `has_many :user_school_infos`.  I'd alias `user.school_info` to the most current UserSchoolInfo.

Reply:
Ah got it. Thanks and sorry for derailing! 😛

Reply:
A cool potential approach to this, as documented in ["Fun with Rails has_one through association:"](https://dev.mikamai.com/2015/03/18/fun-with-rails-hasone-through-association/)

```rb
class User < ApplicationRecord
  has_many :user_school_infos
  has_one :latest_user_school_info,
    class_name: 'UserSchoolInfo',
    conditions: 'end_date is not null',
    order: 'start_date DESC'
  has_one :school_info, through: :latest_user_school_info
end
```

Not sure I've got the syntax exactly right (also derived some from [this SO question](https://stackoverflow.com/questions/9840169/rails-has-one-of-a-has-many-association)) but might be a neat way to replace `:school_info` as a column.

Reply:
@islemaster +1 on deprecating User.school_info_id.  @bethanyaconnor, do you recall if there was a reason that the data team wanted to keep school_info_id in the users table?  I will check with Ben.

Reply:
Found a previous discussion with Ben:

"so we want to keep the most recent version of a teachers `school_info_id` in the `users` table, and then use your new table as a historical log of info teachers have given us. the main reason for keeping the most recent version in `users` is ease of use -- most of the time, we just want to know what a teacher's current school is (which would be accessible via the users table). but the new table will allow us to look back to previous school years. if we didn't, we'd have to be checking every time we run a query to see which row in `user_school_infos` is most recent for each teacher, etc."

Reply:
Let's push back on this. (FYI @bencodeorg)  It seems like this shortcut to make queries _easier_ may also make them incorrect in many cases (any query looking at an extended period of time, for example).  It also assumes a teacher may only have one active school at a time, but even if we enforce that now we should maybe change it in the future.  My dad taught at three schools at once for a while, and I wouldn't be surprised if many of our teachers have a similar situation.

If data really wants a column in users for `school_info_id` maybe we can just do that on the reporting database, or create a view to make working with this data easier in the future.

Reply:
I think both the a) "one active school at a time" assumption and b) using our most recent information on where a teacher teaches, even if out of date are necessary tradeoffs between correctness and simplicity.

If having `school_info_id` in `users` is a major pain, we could definitely implement as a view in Redshift. There would be some work to change anything that uses a user's school info (which is not inconsequential), but definitely possible.

Reply:
It will be great to have a follow-up conversation about this with all the stakeholders.  @islemaster and @bethanyaconnor, I have updated the tests for this PR so it can be merged pending review.  Since the discussion ^ involves removing a column, it will be better if that change is in a different PR.

Reply:
> A cool potential approach to this, as documented in ["Fun with Rails has_one through association:"](https://dev.mikamai.com/2015/03/18/fun-with-rails-hasone-through-association/)
> 
> ```ruby
> class User < ApplicationRecord
>   has_many :user_school_infos
>   has_one :latest_user_school_info,
>     class_name: 'UserSchoolInfo',
>     conditions: 'end_date is not null',
>     order: 'start_date DESC'
>   has_one :school_info, through: :latest_user_school_info
> end
> ```
> 
> Not sure I've got the syntax exactly right (also derived some from [this SO question](https://stackoverflow.com/questions/9840169/rails-has-one-of-a-has-many-association)) but might be a neat way to replace `:school_info` as a column.

This is cool BTW!